### PR TITLE
[BEAM-13416] Introduce Schema provider for AWS model classes extending SdkPojo

### PR DESCRIPTION
**Pull request by mosche**
@echauchot @aromanenko-dev Not really a PR (yet), I was just looking to understand the usage of schemas vs coders better.
The (most) interesting part here is probably `SdkPojoSchema` that works for all the generated AWS models in the SDK v2 (though, in some cases recursive types are used which obviously impose a challenge).

I wasn't able to find a lot of good technical documentation (design decisions) for schemas in Beam, though definitely see the value over low level coders! It seems to make sense to bother users as little as possible with coders and [not expose these on any public API](https://docs.google.com/document/d/1V2FkGGunVgvLwi1dKHr-7mtDuwYjTuvuESV_oPzVnfQ/edit?resourcekey=0-KvfQq-5iCcMlu3f3MFJ-GQ#heading=h.xl97lw3dyot7) (M11). Currently coders are all over the place for AWS IOs, with this it seems feasible to drop them (or deprecate).

Though, I was wondering a couple of things:
- Looks like I cannot register a provider for an interface (`SdkPojo`), that means I have to "announce" it globally using an auto service.
- Is there a way a Sink/ WriteIO can set the coder for it's input like that can be done for a Source on the output? Looks like it's only possible to register such a coder globally, regardless if that sink is used or not.
- I noticed that a RowCoder cannot be automatically inferred for a row despite every row carrying its schema. That makes sense given runners may use the schema ahead of time for query planning. But it made me wonder if there's any relationship between the schema of the row coder and the runtime schema enforced... 

If you have any good resources on the topic, I'd be happy to hear ...

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by mosche** [comment]
@pabloem I remember you made related changes on your recent PR to JdbcIO to add support for generic partitioners. I wonder if you have pointers to documentation or simply advice on the topic :)

**Comment by aromanenko-dev** [comment]
@reuvenlax Could you advise on this please?

**Comment by reuvenlax** [comment]
Can you explain what SdkPojoSchema is? We already have an optimized Pojo schema provider (JavaFieldSchema)

**Comment by mosche** [comment]
@reuvenlax JavaFieldSchema doesn't work for AWS models, the sdk v2 uses immutable objects based on builders / final fields. SdkPojo is the common interface of all of these models.

**Comment by mosche** [comment]
SdkPojoSchema uses the metadata exposed by SdkPojos to read fields and write them via an intermediate builder.

**Comment by reuvenlax** [comment]
JavaFieldSchema should work for immutable objects with builders too. Can we
elucidate what the gap is there?

JavaFieldSchema is fairly optimized using bytecode generation, so it
doesn't have to convert back and forth to a Row object (rather the Row
object simply forwards it accesses to direct accesses of the POJO using
bytecode). If we are able to leverage that, it will be a
significant performance advantage over the SdkPojo schema provider.

**Comment by mosche** [comment]
Thanks, I'll have a closer look

**Comment by mosche** [comment]
@reuvenlax I had a closer look why JavaFieldSchema does not work. There's quite some issues, and this is by far not an exhaustive list:

- Only [public, non-static, non-transient fields](https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/utils/ReflectUtils.java#L129-L151) are considered. Generated models in the AWS SDK are private, however. Using some kind of Field filter in JavaFieldSchema, that can then be overwritten, this could be addressed. But this obviously also means to generate different FieldValueGetter using ByteBuddy.
- Next, more involved I suppose, such AWS SDK models contain final fields only, but there's no known creator. The annotation based approach isn't feasible for such "external" classes.
- An annoyance, but has to be handled as well: These model classes may contain maps with non primitive types. This is not supported. Also, in some cases there might be recursive type hierarchies that can't be supported and have to be dealt with.

Also, I've seen all the optimisations done in JavaFieldSchema and JavaBeanSchema. Honestly, at this point I don't bother much about that yet. I'm more interested to understand what the right approach forward is giving all the AWS IOs currently use hand crafted low level coders.

**Comment by mosche** [comment]
Also, for context, here's a stripped down example how such a `SdkPojo` may look like:
```java
public interface SdkPojo {
    List<SdkField<?>> sdkFields();
}

@Generated("software.amazon.awssdk:codegen")
public final class DeleteMessageRequest extends SqsRequest implements SdkPojo {
    private static final SdkField<String> QUEUE_URL_FIELD = SdkField.<String> builder(MarshallingType.STRING)
            .memberName("QueueUrl")
            .getter(DeleteMessageRequest::queueUrl)
            .setter(Builder::queueUrl)
            .build();

    private static final SdkField<String> RECEIPT_HANDLE_FIELD = SdkField.<String> builder(MarshallingType.STRING)
            .memberName("ReceiptHandle")
            .getter(DeleteMessageRequest::receiptHandle)
            .setter(Builder::receiptHandle)
            .build();

    private static final List<SdkField<?>> SDK_FIELDS = Collections.unmodifiableList(Arrays.asList(QUEUE_URL_FIELD,
            RECEIPT_HANDLE_FIELD));

    private final String queueUrl;

    private final String receiptHandle;

    private DeleteMessageRequest(BuilderImpl builder) {
        super(builder);
        this.queueUrl = builder.queueUrl;
        this.receiptHandle = builder.receiptHandle;
    }

    public final String queueUrl() {
        return queueUrl;
    }

    public final String receiptHandle() {
        return receiptHandle;
    }

    public static Builder builder() {
        return new BuilderImpl();
    }

    public final List<SdkField<?>> sdkFields() {
        return SDK_FIELDS;
    }

    public interface Builder extends SdkPojo {
        Builder queueUrl(String queueUrl);

        Builder receiptHandle(String receiptHandle);
    }

    static final class BuilderImpl implements Builder {
        private String queueUrl;

        private String receiptHandle;

        private BuilderImpl() {}

        public final Builder queueUrl(String queueUrl) {
            this.queueUrl = queueUrl;
            return this;
        }

        public final Builder receiptHandle(String receiptHandle) {
            this.receiptHandle = receiptHandle;
            return this;
        }

        public DeleteMessageRequest build() {
            return new DeleteMessageRequest(this);
        }

        public List<SdkField<?>> sdkFields() {
            return SDK_FIELDS;
        }
    }
}
```

**Comment by reuvenlax** [comment]
On Wed, Mar 2, 2022 at 3:02 AM Moritz Mack ***@***.***> wrote:

> @reuvenlax <https://github.com/reuvenlax> I had a closer look why
> JavaFieldSchema does not work. There's quite some issues, and this is by
> far not an exhaustive list:
>
>    - Only public, non-static, non-transient fields
>    <https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/utils/ReflectUtils.java#L129-L151>
>    are considered. Generated models in the AWS SDK are private, however. Using
>    some kind of Field filter in JavaFieldSchema, that can then be overwritten,
>    this could be addressed. But this obviously also means to generate
>    different FieldValueGetter using ByteBuddy.
>
> But the getter methods are public, right?


>
>    - Next, more involved I suppose, such AWS SDK models contain final
>    fields only, but there's no known creator. The annotation based approach
>    isn't feasible for such "external" classes.
>
> That's something I think could be solved. Yes today we require an
annotation to specify the builder, however we could easily add a
programmatic way to specify this.

>
>    - An annoyance, but has to be handled as well: These model classes may
>    contain maps with non primitive types. This is not supported. Also, in some
>    cases there might be recursive type hierarchies that can't be supported and
>    have to be dealt with.
>
> I think we do support this, or are you referring to the key type?
Recursive type hierarchies are also definitely support.

**Comment by mosche** [comment]
> But the getter methods are public, right?

There's public getters of course, but  not following the "getX" naming convention. 

I've also looked briefly at just extending `GetterBasedSchemaProvider`, that works really well actually. The read part is rather trivial. The field metadata of each SDK model provides an easy way to read fields (without any reflection involved).

```java
static class SdkFieldValueGetter<ObjectT, ValueT> implements FieldValueGetter<ObjectT, ValueT> {
    private final SdkField<ValueT> field;

    public SdkFieldValueGetter(SdkField<ValueT> field) {
      this.field = field;
    }

    @Override
    @Nullable
    public ValueT get(ObjectT object) {
      return field.getValueOrDefault(object);
    }

    @Override
    public String name() {
      return field.memberName();
    }
  }
```

> Recursive type hierarchies are also definitely support.

Problem with recursive types is the immutable `Schema`. I'm not aware of any way to generate such a schema for a recursive hierarchy ... same problem exists for `FieldValueTypeInformation`.
Just verified it quickly, I get a stackoverflow as expected ... so recursive types are definitely not supported.

**Comment by reuvenlax** [comment]
Do you mean mutually recursive? We considered supporting this with schema,
but the complexity was high and there didn't seem to be many good use cases.

**Comment by reuvenlax** [comment]
Unfortunately if you truly need recursive or mutually-recursive types, Beam
Schemas won't work today. We supported nested types, but not true recursion.

**Comment by mosche** [comment]
@reuvenlax Trying to implement this on the existing Schema providers, I kept running into various issues. Also `RowWithGetters` proved to be difficult for this use case. 

I introduce a similar wrapper row delegating to an underlying object very similar to `RowWithGetters` and optimized things a bit. I would be very grateful for your feedback. Though no rush, I'm off for the next week anyways ...

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/values/AwsPojoRow.java [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+    return map.entrySet().stream()
+        .collect(toMap(Map.Entry::getKey, e -> new AwsPojoRow(e.getValue(), schema)));
+  }
+
+  private <T> T cachedValue(int fieldIdx, Function<Field, T> conversion) {
+    Field field = getSchema().getField(fieldIdx);
+    if (valueCache.length == 0) {
+      return conversion.apply(field);
+    }
+    short cacheIdx = field.getOptions().getValueOrDefault(CACHE_IDX_OPTION, (short) -1);
```
Comment (mosche) [inline sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/values/AwsPojoRow.java:149]: @mosche treat `-1` here!

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/values/AwsPojoRow.java [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+import software.amazon.awssdk.core.SdkPojo;
+import software.amazon.awssdk.core.protocol.MarshallingType;
+import software.amazon.awssdk.core.traits.ListTrait;
+import software.amazon.awssdk.core.traits.MapTrait;
+
+/**
+ * A row delegating to an underlying AWS {@link SdkPojo}.
+ *
+ * <p>This is following the example of {@link RowWithGetters}, but optimized for usage with {@link
+ * SdkPojo}s.
```
Comment (TheNeuralBit) [inline sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/values/AwsPojoRow.java:48]: Could you summarize what are the gaps in `RowWithGetters` that you need to address with this implementation? I'm wondering if these are fixes we should apply in  `RowWithGetters` instead of creating an alternate version.

It's worth taking the time to enumerate them - if we really do need to fork `RowWithGetters` in this way, I think we should document what is different and why in this javadoc.

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/values/package-info.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/** Row support for AWS model classes. */
+@Experimental(Kind.SCHEMAS)
+package org.apache.beam.sdk.values;
+
+import org.apache.beam.sdk.annotations.Experimental;
+import org.apache.beam.sdk.annotations.Experimental.Kind;
```
Comment (TheNeuralBit) [inline sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/values/package-info.java:24]: Could we just put `AwsPojoRow` in the `aws2.schemas` package instead?

**Review by TheNeuralBit — COMMENTED** [review]
Hi @mosche thanks for this! I can help review, I have a few initial questions.

FYI @apilloud (relevant for schema proliferation), and @chamikaramj (could enable cross-language AWS IOs)

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/values/AwsPojoRow.java (hunk 2 above)
Comment (mosche) [reply sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/values/AwsPojoRow.java:48]: Thanks so much for looking into this @TheNeuralBit 🙏 

For completeness, I wasn't able to use the existing `JavaFieldSchema` or `JavaBeanSchema` for the following reasons:
- Model fields of the AWS SDK are private (and final).
- Getters / Setters lack expected prefixes to be valid beans.
- (Immutable) instances must be created using a builder (but these obviously lack the expected annotations).
- Some SDK models are self recursive, these are a major annoyance and required logic to eventually skip nested recursive fields as Beam schemas don't support this.

My initial version was implemented using `RowWithGetters`, but it required a couple of work arounds and seemed sub-optimal:

- I faced a mismatch when using logical types and I'm more and more wondering if that's a bug: `RowWithGetters` expects the getter to return the base type, but the target object will contain the input type (unless I misunderstand) forcing a pointless conversion input type -> base type -> input type ([see here](https://github.com/apache/beam/pull/11074#discussion_r818731944)). Particularly when working with collection types that seemed very inefficient.

- The getter factory `Factory<List<FieldValueGetter>>` doesn't work well with AWS model classes. Model instances already contain the necessary metadata (`List<SdkField<?>> sdkFields()`) including getters, making traversal of [nested models / rows](https://github.com/mosche/beam/blob/BEAM-13416-AwsSchemas/sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/values/AwsPojoRow.java#L117-L123) trivial. Using the getter factory, that field metadata isn't easily accessible as [factories operate on the class & schema](https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/values/RowWithGetters.java#L66). 

Additionally, there's a couple of optimisations. `RowWithGetters` might also benefit from these:

- [Nested rows](https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/values/RowWithGetters.java#L111) should be cached, otherwise the cache of such nested rows can't be leveraged

- Handling of collections / maps / iterables can be significantly improved in all cases where `getValue` is the [identity transform](https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/values/RowWithGetters.java#L142). In such cases the original collection / map / iterable can be returned as is (also making caching obsolete in that case).

- Caching using [3 separate hash maps](https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/values/RowWithGetters.java#L57-L59) comes with very high overhead for small rows (3 * `initialCapacity`). I figured I could use a single `Object[]` for caching as it's possible to [precalculate the required cache size and indices](https://github.com/mosche/beam/blob/BEAM-13416-AwsSchemas/sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/schemas/AwsPojoSchema.java#L242-L245) and store this using schema & field options.

File: sdks/java/io/amazon-web-services2/src/test/java/org/apache/beam/sdk/io/aws2/schemas/AwsPojoUtilsTest.java [hunk 4]
```diff
@@ -15,24 +15,22 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.beam.sdk.io.aws2.sqs;
```
[the remark this answers is not in the document]
Comment (mosche) [reply sdks/java/io/amazon-web-services2/src/test/java/org/apache/beam/sdk/io/aws2/schemas/AwsPojoUtilsTest.java:18]: 👀

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/sns/SnsResponseCoder.java [hunk 5]
```diff
@@ -30,7 +30,12 @@
 import org.apache.beam.sdk.coders.VarIntCoder;
 import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.ImmutableList;
 
-/** Custom Coder for WrappedSnsResponse. */
+/**
+ * Custom Coder for WrappedSnsResponse.
+ *
+ * @deprecated Coder of deprecated {@link SnsResponse}.
```
[the remark this answers is not in the document]
Comment (mosche) [reply sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/sns/SnsResponseCoder.java:36]: I was planning to turn on schema coders by default by removing all "legacy" `CoderProviderRegistrar`s and instead adding the general`SchemaProviderRegistrar` for all AWS models. Users can opt-out using schemas using the deprecated `withCoder` setting.

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/values/package-info.java (hunk 3 above)
Comment (mosche) [reply sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/values/package-info.java:24]: I'd certainly like to. Though this is one of the open issues that need discussion: the constructor of `Row` is package private.

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/values/AwsPojoRow.java (hunk 2 above)
Comment (TheNeuralBit) [reply sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/values/AwsPojoRow.java:48]: Thank you this is great!

I think it's fine to build custom SchemaProviders (rather than `JavaFieldSchema` and `JavaBeanSchema`)  for the AWS types. It's a common accepted pattern to build SchemaProviders for different domains (e.g. the protobuf type SchemaProvider in that protobuf extensions). It would be nice if an existing one "just worked", but the bar for creating a new one should be low IMO, as long as we take care to re-use things when appropriate.

> Some SDK models are self recursive, these are a major annoyance and required logic to eventually skip nested recursive fields as Beam schemas don't support this.

Is it OK to just skip those recursive fields?

> I faced a mismatch when using logical types...

Hm yeah this is a frustrating issue that's come up in several other contexts. Let's continue discussing it on the [source PR](https://github.com/apache/beam/pull/11074#discussion_r818731944). It would be nice to find a solution that's generally applicable.

> The getter factory `Factory<List<FieldValueGetter>>` doesn't work well with AWS model classes. Model instances already contain the necessary metadata ...

I'm not quite clear on this one, I guess since I have limited context on AWS model classes vs. Model instances. Is there some reference you could link for this?

> Additionally, there's a couple of optimisations. RowWithGetters might also benefit from these:

Could we separate out these changes and apply them to `RowWithGetters` in a way that's re-usable here?

Comment (mosche) [reply sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/values/AwsPojoRow.java:48]: @TheNeuralBit As suggested I ported some of the changes to `RowWithGetters` https://github.com/apache/beam/pull/17172

Comment (mosche) [reply sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/values/AwsPojoRow.java:48]: Regarding the getter factory, my issue is that it operates on the class level:
```java
getters = fieldValueGetterFactory.create(getterTarget.getClass(), schema);
```

Assuming the hypothetical AWS model class `Message` below, I need a getter factory that operates on the instance level to read `sdkFields()`:

```java
getters = fieldValueGetterFactory.create(getterTarget, schema);
```

Otherwise I have to read that per reflection / or generate byte code to do so. In any case, it introduces a significant in-direction.

```java
public final class Message implements SdkPojo {
   private static final SdkField MESSAGE = SdkField.builder(STRING).name("Message").getter(Message::message).build();
   private static final List<SdkField> SDK_FIELDS = ImmutableList.of(MESSAGE);

  @Override // from SdkPojo 
  public final List<SdkField<?>> sdkFields() { return SDK_FIELDS; }

  private final String message;

  public final String message() { return message; }
}
```

Comment (mosche) [reply sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/values/AwsPojoRow.java:48]: Besides, I'm wondering about the value of the method `name()` in `FieldValueGetter`: It's not used anywhere and prevents using lambdas with `FieldValueGetter`.

**Comment by aromanenko-dev** [comment]
Ping. What is a state of this PR? What are the next steps?

**Comment by TheNeuralBit** [comment]
Hi @mosche does this need a rebase or anything now that #17172 is merged, or is it ready to review?

**Comment by mosche** [comment]
Just wrapping up some other work, then I'll pick this up again 👍 There's a bit of work to be done to rebase this onto master.
Thanks for checking in @TheNeuralBit .

**Comment by mosche** [comment]
@TheNeuralBit I've migrated this PR onto `GetterBasedSchemaProvider` and `RowWithGetters` to take advantage of changes in https://github.com/apache/beam/pull/17172.
I also decided to remove support for self-recursive AWS models (e.g. DynamoDB's `AttributeValue`) for this initial version. Supporting that is a bit problematic with Beam schemas. In such a case a "traditional" coder is still required.

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/schemas/AwsSchemaProvider.java [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+import software.amazon.awssdk.core.SdkPojo;
+import software.amazon.awssdk.utils.builder.SdkBuilder;
+
+/**
+ * Schema provider for AWS {@link SdkPojo} models using the provided field metadata (@see {@link
+ * SdkPojo#sdkFields()}) rather than reflection.
+ *
+ * <p>This provider overrides{@link GetterBasedSchemaProvider#fromRowFunction(TypeDescriptor)} to
+ * instead use the generated builder factories with the setters from {@link SdkField}s, see {@link
+ * SdkField#set(Object, Object)}.
```
Comment (TheNeuralBit) [inline sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/schemas/AwsSchemaProvider.java:60]: This seems like an implementation detail, maybe we should just put it in a (non-javadoc) comment around the overriden method instead?

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/sqs/MessageCoder.java [hunk 7]
```diff
@@ -1,50 +0,0 @@
...
-import java.io.IOException;
-import java.io.InputStream;
-import java.io.OutputStream;
-import java.io.Serializable;
-import org.apache.beam.sdk.coders.AtomicCoder;
-import org.apache.beam.sdk.coders.StringUtf8Coder;
-import software.amazon.awssdk.services.sqs.model.Message;
-
-/** Custom Coder for handling SendMessageRequest for using in Write. */
-public class MessageCoder extends AtomicCoder<Message> implements Serializable {
```
Comment (TheNeuralBit) [inline sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/sqs/MessageCoder.java:29]: This is public and not `@Internal` so technically we shouldn't just remove it outright. That being said this and `SendMessageRequestCoder`, `MessageCoder` are pretty clearly internal coders for use in their respective IOs, so it's probably fine.

I'll leave it up to you - if you want you could mark these coders deprecated and schedule them for removal in a later release.

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/schemas/AwsTypes.java [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+  private static Instant toJodaInstant(Object instant) {
+    return Instant.ofEpochMilli(((java.time.Instant) instant).toEpochMilli());
+  }
+
+  private static SdkBytes toSdkBytes(Object sdkBytes) {
+    return SdkBytes.fromByteArrayUnsafe((byte[]) sdkBytes); // TODO copy or use unsafe?
+  }
+
+  private static byte[] toBytes(Object sdkBytes) {
+    return ((SdkBytes) sdkBytes).asByteArrayUnsafe(); // TODO copy or use unsafe?
```
Comment (TheNeuralBit) [inline sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/schemas/AwsTypes.java:277]: Are these TODOs something we can address now, or should there be an issue tracking it?

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/sns/SnsIO.java [hunk 9]
```diff
@@ -281,22 +281,33 @@ public Write<T> withRetryConfiguration(RetryConfiguration retry) {
     }
 
     /**
-     * Encode the full {@code PublishResult} object, including sdkResponseMetadata and
```
Comment (TheNeuralBit) [inline sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/sns/SnsIO.java:284]: You might use `@link` here instead of `@code`. That way it would have raised an error for the bad reference

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/sns/SnsIO.java [hunk 10]
```diff
@@ -281,22 +281,33 @@ public Write<T> withRetryConfiguration(RetryConfiguration retry) {
...
+     * Encode the full {@code PublishResponse} object, including sdkResponseMetadata and
      * sdkHttpMetadata with the HTTP response headers.
+     *
+     * @deprecated Writes fail fast in case of errors, no need to check headers.
```
Comment (TheNeuralBit) [inline sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/sns/SnsIO.java:287]: - Is this a change in behavior that we need to communicate to users?
- Will these functions be removed later?

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/sns/SnsIO.java [hunk 11]
```diff
@@ -281,22 +281,33 @@ public Write<T> withRetryConfiguration(RetryConfiguration retry) {
...
+    @Deprecated
     public Write<T> withFullPublishResponseWithoutHeaders() {
       return withCoder(PublishResponseCoders.fullPublishResponseWithoutHeaders());
     }
 
-    /** Encode the {@code PublishResult} with the given coder. */
+    /**
+     * Encode the {@code PublishResponse} with the given coder.
+     *
+     * @deprecated Explicit usage of coders is deprecated.
```
Comment (TheNeuralBit) [inline sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/sns/SnsIO.java:308]: nit: mention inferred schemas/AwsSchemaProvider will be used instead.

File: sdks/java/io/amazon-web-services2/src/test/java/org/apache/beam/sdk/io/aws2/schemas/AwsSchemaProviderTest.java [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+
+  private static SdkBytes sdkBytes(String str) {
+    return SdkBytes.fromByteArrayUnsafe(bytes(str));
+  }
+
+  private static byte[] bytes(String str) {
+    return str.getBytes(UTF_8);
+  }
+
+  private <T> Condition<Row> field(String name, T value) {
```
Comment (TheNeuralBit) [inline sdks/java/io/amazon-web-services2/src/test/java/org/apache/beam/sdk/io/aws2/schemas/AwsSchemaProviderTest.java:251]: This is handy, maybe it should be in [SchemaTestUtils](https://github.com/apache/beam/blob/master/sdks/java/core/src/test/java/org/apache/beam/sdk/schemas/utils/SchemaTestUtils.java) for discoverability?

File: sdks/java/io/amazon-web-services2/src/test/java/org/apache/beam/sdk/io/aws2/schemas/AwsSchemaProviderTest.java [hunk 13]
```diff
@@ -0,0 +1,269 @@
...
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.ImmutableMap;
+import org.assertj.core.api.Condition;
+import org.assertj.core.api.SoftAssertions;
+import org.junit.Test;
+import software.amazon.awssdk.core.SdkBytes;
+import software.amazon.awssdk.services.dynamodb.model.AttributeValue;
+import software.amazon.awssdk.services.sqs.model.MessageAttributeValue;
+import software.amazon.awssdk.services.sqs.model.SendMessageRequest;
+
+public class AwsSchemaProviderTest {
```
Comment (TheNeuralBit) [inline sdks/java/io/amazon-web-services2/src/test/java/org/apache/beam/sdk/io/aws2/schemas/AwsSchemaProviderTest.java:49]: Does this exercise all the types in the mapping from `AwsTypes`?

**Review by TheNeuralBit — COMMENTED** [review]
I have a few high-level comments and nits, this looks good overall. I haven't been able to look at the implementation closely though and probably won't be able to before I leave for vacation at the end of today. If you'd like to keep making progress on this while I'm gone maybe @reuvenlax or @lukecwik could take a look.

**Comment by TheNeuralBit** [comment]
retest this please

**Comment by TheNeuralBit** [comment]
Running tests to get a run of the newly forked integration tests (Java_Amazon-Web-Services2_IO_Direct)

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/schemas/AwsSchemaProvider.java [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+ *
+ * <p>Note: Beam doesn't support self-recursive schemas. Some AWS models, such as {@link
+ * software.amazon.awssdk.services.dynamodb.model.AttributeValue} are self-recursive and cannot be
+ * used with this schema provider.
```
[the remark this answers is not in the document]
Comment (mosche) [reply sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/schemas/AwsSchemaProvider.java:64]: The lack of support of self-referential types in Beam was apparently by design, so there's no such issue. My initial idea for AWS was to "fake" it by just skipping over self-referential fields at some depth. But that is obviously not correct and potentially dangerous as it might silently ignore some data.

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/schemas/AwsSchemaProvider.java (hunk 6 above)
Comment (mosche) [reply sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/schemas/AwsSchemaProvider.java:60]: 👍

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/sqs/MessageCoder.java (hunk 7 above)
Comment (mosche) [reply sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/sqs/MessageCoder.java:29]: You're absolutely right pointing this out 👍 Considering `MessageCoder.of()` / `SendMessageRequestCoder.of()` are both package private I think direct removal is justifiable in this case.

File: sdks/java/io/amazon-web-services2/src/test/java/org/apache/beam/sdk/io/aws2/schemas/AwsSchemaProviderTest.java (hunk 12 above)
Comment (mosche) [reply sdks/java/io/amazon-web-services2/src/test/java/org/apache/beam/sdk/io/aws2/schemas/AwsSchemaProviderTest.java:251]: This code depends on `assertj-core` and probably doesn't fit into SchemaTestUtils for that reason.
I could rewrite it as `Matcher` if you think it's worth it,e.g.:
```java
public static Matcher<Row> containsFieldValues(Map<String, Object> expectedFieldValues) 
```

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/schemas/AwsTypes.java (hunk 8 above)
Comment (mosche) [reply sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/schemas/AwsTypes.java:277]: I had a look at what GetterBasedSchemaProvider does for `byte[]` / `ByteBuffer`. It's exposing the mutable bytes as well, so I'll keep the unsafe operations here. 👍

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/sns/SnsIO.java (hunk 10 above)
Comment (mosche) [reply sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/sns/SnsIO.java:287]: This is just a clarification why there's no need for a coder that includes HTTP response metadata, there's no change in behavior.
Plans is to remove these after a few releases.

File: sdks/java/io/amazon-web-services2/src/test/java/org/apache/beam/sdk/io/aws2/schemas/AwsSchemaProviderTest.java (hunk 13 above)
Comment (mosche) [reply sdks/java/io/amazon-web-services2/src/test/java/org/apache/beam/sdk/io/aws2/schemas/AwsSchemaProviderTest.java:49]: Done ✅

**Comment by mosche** [comment]
Thanks a lot for your review @TheNeuralBit 🙏 
Enjoy your time off!

**Comment by mosche** [comment]
@TheNeuralBit Could you take a final look, I just rebased to account for the byte buddy changes on master ...

**Review by TheNeuralBit — APPROVED** [review]
LGTM, thank you!

You might consider creating an issue for removing the deprecated functionality, and linking it to a future release milestone. That will help us remember to actually remove it.

**Comment by mosche** [comment]
Thanks a lot @TheNeuralBit :)
